src/neo4j_graphrag/experimental/components/entity_relation_extractor.py

```diff
@@ -337,3 +344,147 @@ async def run(
+
+        Remove invalid entities,relationships, and properties.
+        If an entity is removed, all of its relationships are also removed.
+        If no valid properties remain for an entity, remove that entity.
+        """
+        # enforce nodes (remove invalid labels, strip invalid properties)
+        filtered_nodes = self._enforce_nodes(graph.nodes, schema)
+
+        # enforce relationships (remove those referencing invalid nodes or with invalid
+        # types or with start/end nodes not conforming to the schema, and strip invalid
+        # properties)
+        filtered_rels = self._enforce_relationships(
+            graph.relationships, filtered_nodes, schema
+        )
+
+        return Neo4jGraph(nodes=filtered_nodes, relationships=filtered_rels)
+
+    def _enforce_nodes(
+            self,
+            extracted_nodes: List[Neo4jNode],
+            schema: SchemaConfig
+    ) -> List[Neo4jNode]:
+        """
+            Filter extracted nodes to be conformant to the schema.
+
+            Keep only those whose label is in schema.
+            For each valid node, filter out properties not present in the schema.
+            Remove a node if it ends up with no valid properties.
+            """
+        valid_nodes = []
+        if self.enforce_schema == SchemaEnforcementMode.STRICT:
+            for node in extracted_nodes:
+                if node.label in schema.entities:
+                    schema_entity = schema.entities[node.label]
+                    filtered_props = self._enforce_properties(node.properties,
+                                                              schema_entity)
+                    if filtered_props:
+                        # keep node only if it has at least one valid property
+                        new_node = Neo4jNode(
+                            id=node.id,
+                            label=node.label,
+                            properties=filtered_props,
+                            embedding_properties=node.embedding_properties,
+                        )
+                        valid_nodes.append(new_node)
+        # elif self.enforce_schema == SchemaEnforcementMode.OPEN:
+            # future logic
+        return valid_nodes
+
+    def _enforce_relationships(
+            self,
+            extracted_relationships: List[Neo4jRelationship],
+            filtered_nodes: List[Neo4jNode],
+            schema: SchemaConfig
+    ) -> List[Neo4jRelationship]:
+        """
+        Filter extracted nodes to be conformant to the schema.
+
+        Keep only those whose types are in schema, start/end node conform to schema,
+        and start/end nodes are in filtered nodes (i.e., kept after node enforcement).
+        For each valid relationship, filter out properties not present in the schema.
+        """
+        valid_rels = []
+        if self.enforce_schema == SchemaEnforcementMode.STRICT:
+            valid_node_ids = {node.id for node in filtered_nodes}
```

Comment:
Maybe create a lookup dict `{node_id: node_label}` here since we're already looping over filtered nodes, and get rid of the `_get_node_label` method.